docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}");
+    });
+
+```
+
+After creation, output nodes must be then registered with the pipeline context:
+
+```csharp
+    context.RegisterOutput(output);
+```
+
+A generator may create and register multiple output nodes as part of the
+pipeline, but an output cannot be further transformed once it is created.
+
+It is expected that there will be equivalent `Generate...` methods for
+generating non source outputs such as `GenerateArtifact(...)` or
+`GenerateEmbeddedFile(...)`. Splitting the outputs by type produced allows the
+host (such as the IDE) to not run outputs that are not required. For instance
+artifacts are only produced as part of a command line build, so the IDE has no
+need to run the artifact based outputs or steps that feed into it.
+
+**OPEN QUESTION:** Should we just make the registration automatic, as part of
+calling `.GenerateSource`? Its trickier to plumb through behind the scenes, but
+I don't see any obvious use case for an author to create a node and not register
+it? I've footgunned myself already with this where I forgot to register the
+output and wondered why nothing was happening.
+
+**OPEN QUESTION** Should we have `GenerateBatch...` versions of the output
+methods? This can already be achieved by the author calling `BatchTransform`
+before calling `Generate...` so would just be a helper, but seems common enough
+that it could be useful.
+
+### Simple example
+
+Putting together the various steps outlined above, an example incremental
+generator might look like the following:
+
+```csharp
+[Generator(LanguageNames.CSharp)]
+public class MyGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext initContext)
+    {
+        initContext.RegisterExecutionPipeline((context) =>
+        {
+            // get the additional text source
+            var additionalTexts = context.Sources.AdditionalTexts;
+
+            // apply a 1-to-1 transform on each text, extracting the path
+            var transformed = additionalTexts.Transform((text) => text.Path);
+
+            // batch the collected file paths into a single collection
+            var batched = transformed.BatchTransform(texts => texts);
+
+            // create an output that takes the file paths from the above batch and makes some user visible syntax
+            var output = batched.GenerateSource((sourceContext, filePaths) =>
+            {
+                sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}
+");
+            });
+
+            // register the output context.RegisterOutput(output); }); } }
+```
+
```

Comment:
Think you need an example which shows usage of a `SemanticModel` and `SourceText` here. That is one of the most common forms of generators, basically looking for an attribute trigger. Customers going from V1 to V2 are going to want to see that type of example that they can look to for guidance.

Reply:
Yeah thats the 'syntax nodes' bit down below. The problem is I'm still trying to figure out exactly what that actually *looks* like. We have some ideas, but still trying to find the final shape for it :/

Reply:
Just read through this and I don't have the context so I might have skipped something, but this is the biggest question for me as well.

Basically, I need to know 

1) Was this type annotated with this particular attribute
2) Did any fields/properties change in the type

(2) may be easy to do via a simple overly-aggressive "is this text the same", (1) I might be able to cheat and just ask if the name of the attribute is the same, but theoretically the binding for the attribute could be different. Am I supposed to care? And if I do grab the symbol and ask if two symbols are the same, presumably that's not valid because you can't compare symbols from two compilations. So what do I do? Ask if they're the same fully-qualified name?

It would be good to spell this out (especially the part about not being able to compare symbols).

Reply:
This is also my main question.  Without the SemanticModel being part of the input (and ideally portions of it cached), I'm not sure this would give us much savings in our use of generators.

Ideally, this would work as follows:
1. Even if input syntax does not change, but one of the dependent symbols looked-up did, the pipeline would re-execute.
2. If input syntax does not change, and all dependent symbols stay the same; the previous output would be re-used.